packages/flutter/lib/src/material/text_field.dart

```diff
@@ -1302,30 +1302,33 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
       semanticsMaxValueLength = null;
     }
 
-    return MouseRegion(
-      cursor: effectiveMouseCursor,
-      onEnter: (PointerEnterEvent event) => _handleHover(true),
-      onExit: (PointerExitEvent event) => _handleHover(false),
-      child: IgnorePointer(
-        ignoring: !_isEnabled,
-        child: AnimatedBuilder(
-          animation: controller, // changes the _currentLength
-          builder: (BuildContext context, Widget? child) {
-            return Semantics(
-              maxValueLength: semanticsMaxValueLength,
-              currentValueLength: _currentLength,
-              onTap: widget.readOnly ? null : () {
-                if (!_effectiveController.selection.isValid)
-                  _effectiveController.selection = TextSelection.collapsed(offset: _effectiveController.text.length);
-                _requestKeyboard();
-              },
-              onDidGainAccessibilityFocus: handleDidGainAccessibilityFocus,
+    return FocusArea(
+      focusNode: focusNode,
```

Comment:
This is a little odd. It's usually not OK to give the same focus node to two widgets: it messes with where the focus node ends up in the focus tree, and can cause some really weird focus issues, since the where it ends up depends on who builds last.

In this case, since you're just listening to the focus node, it's probably OK, but we should probably not call this attribute `focusNode`, so that we differentiate it from all the other widgets that take a focus node in order to use it to implement focus.  Maybe `watchedFocusNode`, `monitoredFocusNode`? (don't love either of those, really)

I do worry about lifecycle issues: if this widget is listening to the focus node, I think it will assert if it tries to remove listener on that node if the node has already been disposed by the widget that owns it.

Reply:
yes, this one is a head scratcher. The problem I have is that at the render object layer, the focus not only belongs to a tiny part of the text field. But if you look at what is on screen, the focus is clearly attributable to a much wider area that encompasses the decoration / borders / et cetera.



> It's usually not OK to give the same focus node to two widgets

i'm not really giving the focus in this case, more like using it as a key/token. I can definitely rename to something else, but overall I'd love for there to be a cleaner solution here

Reply:
I am also curious, if user click the decoration, will that bring the focus to the editable text? if so, how does it know it hits the editable text?

Reply:
This is already the behavior today.

What this change does is makes sure that we do not remove focus if the user clicks the decoration. Doing so would cause it to immediately be re-acquired by the text field's gesture detection.

Reply:
I wonder if we can make focusnode attach to the TextFieldState instead of EditableTextState?

Reply:
I think the editable text state uses the focus too though, which goes back to @gspencergoog 's point about not re-using a focus object (though we kind of already are?)

Reply:
we could conditionally attach the focus to either state based on how the widget was created. Seems iffy TBH

Reply:
You can attach in TextFieldState and pass into the EditableText, and the EditableText just listen to the focusnode. I think as long as the EditableTextState dispose before TextFieldState dispose, there should be any issue. @gspencergoog may know more. This way, the focusNode.currentContext.findRenderObject will return the renderobject of the TextDecoration.

Reply:
let me play around with this a bit, I'm not sure if EditableText needs the focus attachment if the text field already has it

Reply:
flutter gallery seems to work OK, I don't dare run the tests...